### PR TITLE
Prohibit automatic downgrades

### LINKED ISSUE #13107: Add DB-Schema-Version Check on Gitea start
make sure the db was not already migrated to a newer version and the user has switched back to a old version without restoring the state before migrations took in place

### PR DESCRIPTION
**Pull request by 6543**
close #13107

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
@lafriks added print to stderr too

while testing on backport ... I noticed it wont show up in the terminal - hope It's still fine

File: models/migrations/migrations.go [hunk 1]
```diff
@@ -315,12 +316,16 @@ Please try upgrading to a lower version first (suggested v1.6.4), then upgrade t
 		return nil
 	}
 
+	// Downgraded Gitea not supported
 	if int(v-minDBVersion) > len(migrations) {
-		// User downgraded Gitea.
-		currentVersion.Version = int64(len(migrations) + minDBVersion)
-		_, err = x.ID(1).Update(currentVersion)
-		return err
+		msg := fmt.Sprintf("Downgrade Gitea from '%d' to '%d' not supported.\nIf you really have to execute `UPDATE version SET version=%d WHERE id=1;`\n",
```
Comment (6543) [inline models/migrations/migrations.go]:
```suggestion
		msg := fmt.Sprintf("Downgrade Gitea from '%d' to '%d' not supported.\nIf you really have to, execute `UPDATE version SET version=%d WHERE id=1;`\n",
```

**Comment by 6543** [comment]
backport  #13111

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Or we can add a flag or an environment to force downgrade, but not update the database manually.

**Comment by 6543** [comment]
@lunny did had a quick look, a flag would be a nice optional enhancement, but we have to refactor some code then, so migration functions can have options then ...

**Comment by 6543** [comment]
:rocket:

File: models/migrations/migrations.go [hunk 2]
```diff
@@ -315,12 +316,16 @@ Please try upgrading to a lower version first (suggested v1.6.4), then upgrade t
 		return nil
 	}
 
+	// Downgraded Gitea not supported
 	if int(v-minDBVersion) > len(migrations) {
-		// User downgraded Gitea.
-		currentVersion.Version = int64(len(migrations) + minDBVersion)
-		_, err = x.ID(1).Update(currentVersion)
-		return err
+		msg := fmt.Sprintf("Downgrade Gitea from '%d' to '%d' not supported.\nIf you really have to, execute `UPDATE version SET version=%d WHERE id=1;`\n",
```
Comment (CirnoT) [inline models/migrations/migrations.go]:
```suggestion
		msg := fmt.Sprintf("Downgrading Gitea from '%d' to '%d' is not supported and may result in loss of data integrity.\nIf you really know what you're doing, execute `UPDATE version SET version=%d WHERE id=1;`\n",
```
Maybe something like this?

**Comment by 6543** [comment]
:rocket:

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -315,12 +316,16 @@ Please try upgrading to a lower version first (suggested v1.6.4), then upgrade t
 		return nil
 	}
 
+	// Downgraded Gitea not supported
 	if int(v-minDBVersion) > len(migrations) {
-		// User downgraded Gitea.
-		currentVersion.Version = int64(len(migrations) + minDBVersion)
-		_, err = x.ID(1).Update(currentVersion)
-		return err
+		msg := fmt.Sprintf("Downgrading Gitea from '%d' to '%d' is not supported and may result in loss of data integrity.\nIf you really know what you're doing, execute `UPDATE version SET version=%d WHERE id=1;`\n",
```
Comment (zeripath) [inline models/migrations/migrations.go:321]: The version number here is the Database version not the Gitea version.

**Pull request merged**